Add pre-commit setup

jooola opened the pull request:
- Add pre-commit hooks
- Add github action to enforce pre-commit setup

For any hooks required for a 'sub project', for
instance the UI. It will be possible to create custom hooks,
and call some package scripts in the package.json file.

Fixes #1208

paddatrapper commented:
I'm going to cherry-pick 0d711a7067cf6520093ee1ba9cc57fa1db992981 (unparsable Python file) into master now, as it means that master is currently broken and un-deployable

paddatrapper commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,14 @@
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+      - uses: pre-commit/action@v2
```
```suggestion
      - uses: pre-commit/action@v2.0.2
```

jooola replied:
Ok, I guess dependabot will have a bit more work to do.

As reference : https://github.com/pre-commit/action/issues/104#issuecomment-843821467

jooola commented:
Ok, I done for today, I'll continue tomorrow.

These changes are already massive and I would recommend to review commit per commit as I trying to make change by commit hook.

Some changes might be a bit too bold, and I'm not entirely sure of some changes.

I think we should merge this when it is ready and I will re enable and fix pylint in another PR as it might need some code changing.

jooola commented:
Ok, I rebased, and cleaned some extra stuff. Some improvement will still be needed in the future.

Here is what I feel about this commit list:

- Add pre-commit setup (Ok)
- Fix missing encoding pragma on python files (Ok, python3 uses it by default, so if we don't rely python 2 and we are sure about writing utf8 only code, depends on editors, this can be removed)
- Fix spelling (Ok)
- Format code using prettier (Ok)
- Format code using black  (Ok)
- Pin pre-commit action version  (Ok)
- Bump pre-commit/actions version to 2.0.3  (Ok)
- Fix trailing whitespaces in files  (Ok)
- Fix end of files (Ok)
- Fix executable bit  (Not totally sure if everything is good in this change)
- Set end of line to LF
- Fix shebangs (Same as 'Fix executable bit', maybe I break some installation script, but seems unlikely)
- Fix mixed line ending
- More spelling fix (dangerous) (These spelling changes are code, so it seems a good fix, but someone need to make sure of it)
- Remove broken symlink (Maybe it is needed after a composer install ?)
- Fix libretime-api (Ok)
- temp: xDisable pylint (Ok, will send another PR for this, as it is a lot of work)
- Rename tests to test in airtime*analyzer (Ok, didn't try to run tests though)
- Rename test*_ to _\_test in api_clients (Ok, didn't try to run tests though)
- Ignore python test naming for libretime api (OK)

I didn't add some other features, such as formatting some css or js files as it will mostly increase the PR size and this can be done later one. I also found many minified css/js files and didn't want to format them. Using a package manager should be the way to go, but that's not in the scope of this change.

This PR is ready to be reviewed.

jooola commented:
I also didn't find any place to document the use of pre-commit, where should such document go ? CONTRIBUTE.md seems the file, but there is no development document in there.

paddatrapper commented:
Pragmas aren't really necessary, as we are python 3 only, utf-8. But I don't see the harm in adding them.

Spelling changes all good fine too

jooola commented:
I think having a cleaner code is better, so I am for removing them. I wasn't sure about the minimal python version used and if we had any specific use case for this pragma.

paddatrapper reviewed (commented):
Looks good aside from the one change. The spelling corrections are good. Once #1223 is merged, please rebase onto master so that we can run python and PHP tests to verify there are no unintended consequences. I didn't see any in my review, but machines are far better at noticing than me.

The developer documentation is probably best added to `docs/_docs/contribute.md` and `CONTRIBUTING.md`.

jooola commented on docs/_docs/playout-history.md:
```diff
@@ -32,14 +32,13 @@ The **History Templates** page on the History menu enables you to prepare report
 ![](/img/new-hist-temp.png)
 
-Either of these actions opens a page in which you can name the new template, and add or remove elements from the list on the left. To add a new element from the list on the right, click the plus icon for the item you require. If the element you require is not listed, you can use the **Add New Field** box at the lower end of the right side column. Select *string*, *boolean*, *integer*, or *float*, ...
+Either of these actions opens a page in which you can name the new template, and add or remove elements from the list on the left. To add a new element from the list on the right, click the plus icon for the item you require. If the element you require is not listed, you can use the **Add New Field** box at the lower end of the right side column. Select _string_, _boolean_, *integer*, or _float_, ...
```
Ok, this is a nice one, I found a bunch of weird space chars here and there:

![image](https://user-images.githubusercontent.com/19195485/120172364-40a24980-c203-11eb-978e-4a0609d1539c.png)

After replacement, everything works as expected (prettier didn't change it because of that char).

jooola commented:
Ok I added some extra commits to replace some weird chars. Nothing big. And fixed this `*integer*` thingy.


Tell when and if I should squash this PR or if you do it.

paddatrapper commented:
@jooola #1223 has been merged, can you please rebase on master now?

paddatrapper commented:
The merge conflict is from #1225

jooola commented:
@paddatrapper Ok, just rebased and merged conflict.

paddatrapper commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,14 @@
+name: lint
```
```suggestion
name: Lint
```

paddatrapper replied:
Just to match the other actions

jooola replied:
Ok, gonna fix this.

Also, I think this is gonna change any way. I have plans to run pylint and pytest in the same workflow, to reduce setup/teardown, work with a large python matrix, and only test python apps. But the pytest section is coming with the nose => pytest migration for analyzer.

paddatrapper commented:
@Robbt What are your thoughts here? I think this is a good change, but don't want to do it without some consensus.

@jooola can you please add some documentation to `CONTRIBUTING` and `docs/_docs/contribute.md` about how to get this set up in a dev environment?

jooola commented:
Hey sorry for this last minute change, I also added isort pre-commit hook. It sort imports in python files.

paddatrapper commented:
I'm going to wait until Monday for feedback. If nothing is received by then, I'll merge

jooola commented:
Going to solve the merge conflict again
